alfred-telemetry-platform/src/main/java/eu/xenit/alfred/telemetry/binder/solr/sharding/SolrShardingMetrics.java

```diff
@@ -48,18 +49,34 @@ public void updateMetrics() {
                 Tags shardTags = flocTags.and(Tags.of("shard", String.valueOf(shard.getInstance())));
                 Set<ShardInstance> shardInstances = solrShardingMetricsContainer.getShardInstances(shard);
                 int numberOfActiveShardInstances = (int) shardInstances.stream().filter(shardInstance ->
-                        solrShardingMetricsContainer.getReplicaState(shardInstance) == ReplicaState.ACTIVE).count();
+                        solrShardingMetricsContainer.getReplicaState(shardInstance).isPresent()).filter(shardInstance ->
+                        solrShardingMetricsContainer.getReplicaState(shardInstance).get() == ReplicaState.ACTIVE).count();
                 setAndCreateMetricIfNotExists("activeShardInstances", numberOfActiveShardInstances,
                         shardTags, "number");
                 shardInstances.forEach(shardInstance -> {
                     Tags instanceTags = shardTags.and(Tags.of("instanceHost", shardInstance.getHostName()));
-                    ShardState shardState = solrShardingMetricsContainer.getShardState(shardInstance);
+                    ShardState shardState = solrShardingMetricsContainer.getShardState(shardInstance).orElseGet(() -> {
+                        ShardState state = new ShardState();
+                        state = new ShardState();
+                        state.setLastIndexedChangeSetId(-1);
+                        state.setLastIndexedTxId(-1);
+                        state.setLastIndexedChangeSetCommitTime(-1);
+                        state.setLastIndexedTxCommitTime(-1);
+                        state.setMaster(false);
+                        state.setLastUpdated(-1);
+                        return state;
+                    });
                     setAndCreateMetricIfNotExists("lastIndexedChangeSetId",
                             shardState.getLastIndexedChangeSetId(), instanceTags, "number");
                     setAndCreateMetricIfNotExists("lastIndexedTxId", shardState.getLastIndexedTxId(),
                             instanceTags, "number");
-                    setAndCreateMetricIfNotExists("instanceMode",
-                            solrShardingMetricsContainer.getReplicaState(shardInstance).ordinal(), instanceTags, "enumValue");
+                    Optional<ReplicaState> replicaStateOpt = solrShardingMetricsContainer.getReplicaState(shardInstance);
+                    //instance mode 0 means active, 1 means lagging and 2 means silent.
```

Comment:
Did you event "`2` means silent" or is that an official value ?

If not, `-1` could be more appropriate ?

Reply:
It's an official value. If a shard is silent, alfresco returns 2 as instance mode already.

But I could return -1 if you think that's better to show that that shard is non-existent.

Reply:
Silent does not mean non-existent. It means that the solr node has not contacted Alfresco for some time. It may be crashed or not able to contact Alfresco, which is a different state from non-existent.

Reply:
I'll put -1 in then as value.

Reply:
You can use [`Optional.orElse()`](https://docs.oracle.com/javase/8/docs/api/java/util/Optional.html#orElse-T-) to get rid of the `isEmpty()` check below.

I think something like this would work to replace everything else below in a more compact and functional style.

```java
int instanceMode = solrShardingMetricsContainer.getReplicaState(shardInstance).map(ReplicaState::ordinal).orElse(-1);
```